test/test_delaunay.cpp

```diff
@@ -0,0 +1,30 @@
+#include <gtest/gtest.h>
+
+extern "C" {
+#include <math.h>
+
+#include "debug.h"
+#include "delaunay.h"
+}
+
+TEST(test_delaunay, test_delaunay_reduce_layer) {
+    double min_lattice[3][3];
```

Comment:
I'm always paranoid with C code. Fo you need to explicitly deallocate these? I prefer to use `std::vector` to be safe and pass c_pointer to C functions and use reinterpret_cast when needed

Reply:
Do you comment in general? Or some specific point?

Reply:
I see your point. It is uneasy to take a balance of feature developments and fixing holes. I want to merge this PR without going far from the original aim of PR.

Reply:
I understand, but at the very least new code should not introduce new holes. Whenever you are touching a part of a code that your IDE/clang-tidy and any other tools tells you to fix, then do a quick fix for that, otherwise the holes are never addressed.

Reply:
Can `invalid use of pointers to arrays with different qualifiers in ISO C before C2X [-Wpedantic]` and `pointers to arrays with different qualifiers are incompatible in ISO C [-Wpedantic]`
` be postponed to fix? Or just removing `const` can be a solution?

Reply:
I did a bit of experimentation and research and tl;dr: we don't need to address "invalid * with different qualifiers".
There aren't concerning warnings in this PR now and I did a `clang-tidy` test 
```console
$ pre-commit run --hook-stage manual clang-tidy --from-ref origin/develop --to-ref lan496/layer-group-refactor
clang-tidy...............................................................Passed
```
All seem well for this PR at least on this front. There are still the conversations that I have not marked as resolved which should be addressed and I'm all :+1: for now.

<details><summary>Actually concerning warnings</summary>

The actual concerning warnings are not in this
```
src/refinement.c:1451:9: warning: 'mat_multiply_matrix_di3' accessing 72 bytes in a region of size 24
src/spacegroup.c:1150:51: warning: 'j' may be used uninitialized
```

</details>
<details><summary>About "invalid * with different qualifiers"</summary>

According to the wording in the C standard pre C23, it is interpreted that:
```c
const double matrix[2][3]
```
which is a pointer to a (double-typed) pointer has the const such that it is a pointer to a (double-typed) const pointer, instead of the more natural pointer to a (const double-typed) pointer, meaning that in the former you cannot change where a slice is pointing to and in the latter you cannot change the values of the matrix array.

But in reality no compiler is following the logic of the former and instead they all follow the latter which is consistent with C++ standard as well. And in C23 they have changed the wording so that it is correctly interpreted as the latter (values are constant, and not pointers to the array slice is constant). Indeed if you set `CMAKE_C_STANDARD=23` these warnings go away.

Here is a more detailed example of my tests:
```c
#include <stdio.h>

void test_const_matrix(const int matrix[][2]){
    int vector[2] = {2,3};
//    matrix[0][0] = 2;
    matrix = vector;
    matrix = &vector;
}
void test_const_vector(const int vector[2]){
//    vector[0] = 2;
}
void test_matrix(int matrix[][2]){
    int vector[2] = {2,3};
    matrix[0][0] = 2;
    matrix = vector;
    matrix = &vector;
}
void test_vector(int vector[2]){
    vector[0] = 2;
}

int main() {
    int t_matrix[2][2] = {0};
    int t_vector[2] = {0};
    test_const_matrix(t_matrix);
    test_const_vector(t_vector);
    test_matrix(t_matrix);
    test_vector(t_vector);
    printf("Hello, World!\n");
    return 0;
}
```
Here `test_const_matrix` would fail as expected with `matrix[0][0]=2` even though the wording would indicate otherwise even when I bump down the standard. Although I did not test with antiquated `gcc` compilers to confirm this behaviour is true in old compilers as well. (Note that in the example given there are additional warnings about invalid operations that should be considered).

</details>

I'll leave this conversation unresolved for now, just for visibility.

Reply:
Thanks @LecrisUT.

For `warning: 'mat_multiply_matrix_di3' accessing 72 bytes in a region of size 24`, I have really no idea.

For `warning: 'j' may be used uninitialized`, this is OK, which is confirmed by assert.

Reply:
> For `warning: 'j' may be used uninitialized`, this is OK, which is confirmed by assert.

Not really, I will explain in a PR  where I address the clang-tidy issues.